### PR TITLE
Orchestrator zone get api

### PR DESCRIPTION
**Pull request by johnson-jeb**
Create new module for zone-service and create configurations required
for new module. Update settings.gradle to include the new module and
directory.

Add dashboard uri required in the zone response to uaa.yml for main and
integration_test_properties.yml for test. This property has to configure
for each environment.

Create controller class for zone and add get method based on
documentation [1]
Add validation and exceptions and customize the error message specific
for orchestrator zone contract.
Create a new service for zone and reuse the DAO class from identity zone
. Write a new method in identity zone dao class to fetch identity zone
by name. Reuse the exiting query and add condition for name and active.
Build the response based on the format specified in the document [1].

Update spring security config to include orchestrator zone get API. Copy
the configuration from identity zone and use it for orchestrator zone.

Create model class based on the structure defined in the confluence [1].
Add lombok annotations for getter/setter and constructor.

Add unit tests for ZoneService using Mockito and cover all the scenario.

Add Mockmvc tests for ZoneController using Spring mock mvc. Copy the
configurations from existing test. Cover all the scenario and
validations.

Add Integration tests for ZoneController and copy the required
configuration from Identity zone integrations tests.

Also write the test cases for newly written DAO class.

[1] https://devcloud.swcoe.ge.com/devspace/display/RMOCM/Orchestrator+UAA+Zone+provisioning+API#OrchestratorUAAZoneprovisioningAPI-GETAPI

### CODE REVIEW

File: uaa/src/main/webapp/WEB-INF/spring/multitenant-endpoints.xml [hunk 1]
```diff
@@ -20,6 +20,20 @@
           destroy-method="reset">
     </bean>
 
+    <http name="orchestratorZoneSecurity" pattern="/zones/**" create-session="stateless"
+          entry-point-ref="oauthAuthenticationEntryPoint"
+          use-expressions="true" authentication-manager-ref="emptyAuthenticationManager"
+          xmlns="http://www.springframework.org/schema/security">
+
+        <intercept-url pattern="/zones" access="#oauth2.hasScopeInAuthZone('zones.read') or #oauth2.hasScopeInAuthZone('zones.{zone.id}.admin') or #oauth2.hasScope('zones.write')" method="GET"/>
```
Comment (Yogeshkumarb57) [inline uaa/src/main/webapp/WEB-INF/spring/multitenant-endpoints.xml:28]: pattern="/zones" for GET method, will it allow name query param?

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/ConnectionDetails.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package org.cloudfoundry.identity.uaa.zone.model;
+
+import lombok.Data;
+
+@Data
+public class ConnectionDetails {
```
Comment (Yogeshkumarb57) [inline zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/ConnectionDetails.java:6]: Not sure how, but you used all these model classes in previous commit but creating them here in this commit?

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/ZoneControllerMockMvcTests.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+    @Test
+    void testGetZone() throws Exception {
+        //TODO: remove this createIdentityZone method once the create orchestrator zone API is implemented and use it to create zone.
+        IdentityZone identityZone = createIdentityZone();
+        MvcResult result = mockMvc.perform(
+                   get("/zones").param("name", "test-name")
+                                .header("Authorization", "Bearer " + identityClientZonesReadToken))
+               .andExpect(status().isAccepted()).andReturn();
+        if (HttpStatus.OK.is2xxSuccessful()) {
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/ZoneControllerMockMvcTests.java:149]: Why if condition? for successful GET, it should always validate response.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/ZoneControllerMockMvcTests.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        identityZone.getConfig().getSamlConfig().setCertificate(serviceProviderCertificate);
+        MvcResult result = mockMvc.perform(
+                                      post("/identity-zones")
+                                          .header("Authorization", "Bearer " + uaaAdminClientToken)
+                                          .contentType(APPLICATION_JSON)
+                                          .content(JsonUtils.writeValueAsString(identityZone)))
+                                  .andExpect(status().is(HttpStatus.CREATED.value()))
+                                  .andReturn();
+
+        if (HttpStatus.OK.is2xxSuccessful()) {
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/ZoneControllerMockMvcTests.java:200]: Same here, Why if condition? if it is creating zone, then i guess we can remove if check

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/ConnectionDetails.java (hunk 2 above)
Comment (johnson-jeb) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/model/ConnectionDetails.java:6]: Rebased and ordered the commit.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/ZoneControllerMockMvcTests.java (hunk 3 above)
Comment (johnson-jeb) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/ZoneControllerMockMvcTests.java]: Removed.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/ZoneControllerMockMvcTests.java (hunk 4 above)
Comment (johnson-jeb) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/mock/zones/ZoneControllerMockMvcTests.java]: Removed.

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneController.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        return new ResponseEntity<>("{\"message\",\"Request failed due to a validation error.\" }", BAD_REQUEST);
+    }
+
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<String> handleValidationException(MethodArgumentNotValidException e) {
+        return new ResponseEntity<>("{\"message\", \""+ e.getMessage() +"\" }", BAD_REQUEST);
+    }
+
+    @ExceptionHandler(MissingServletRequestParameterException.class)
+    public ResponseEntity<String> handleConstraintViolationException(MissingServletRequestParameterException e) {
```
Comment (Yogeshkumarb57) [inline zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneController.java:60]: Minor naming mistake, same method name handleConstraintViolationException() is used in lines 60 and 65.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+
+    @Test
+    public void testGetZone() {
+        //TODO: delete once the orchestrator create API implemented
+        IdentityZone identityZone = createZone();
+        assertNotNull(identityZone);
+        ResponseEntity<ZoneResponse> response = client.getForEntity(
+            serverRunning.getUrl("/zones") + "?name=" + identityZone.getName(),
+            ZoneResponse.class);
+        if (response.getStatusCode().is2xxSuccessful()) {
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java:72]: Please remove if condition from here also like we did for mock movc tests.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            assertEquals(identityZone.getId(), zoneResponse.getConnectionDetails().getZone().getHttpHeaderValue());
+        }
+    }
+
+    @Test
+    public void testGetZone_Notfound() {
+        ResponseEntity<String> response = client.getForEntity(
+            serverRunning.getUrl("/zones") + "?name=random-name",
+            String.class);
+        if (response.getStatusCode().is4xxClientError()) {
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java:93]: Please remove if condition from here also like we did for mock movc tests.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            assertEquals("{\"message\", \"Zone[random-name] not found.\" }", response.getBody().toString());
+        }
+    }
+
+    @Test
+    public void testGetZone_EmptyError() {
+        ResponseEntity<String> response = client.getForEntity(
+            serverRunning.getUrl("/zones"),
+            String.class);
+        if (response.getStatusCode().is4xxClientError()) {
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java:105]: Please remove if condition from here also like we did for mock movc tests.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+                response.getBody().toString());
+        }
+    }
+
+    @Test
+    public void testGetZone_NameRequiredError() {
+        ResponseEntity<String> response = client.getForEntity(
+            serverRunning.getUrl("/zones") + "?name=",
+            String.class);
+        if (response.getStatusCode().is4xxClientError()) {
```
Comment (Yogeshkumarb57) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java:119]: Please remove if condition from here also like we did for mock movc tests.

**Review by Yogeshkumarb57 — CHANGES_REQUESTED** [review]
@johnson-jeb Review, please address my comments.

File: uaa/src/main/resources/uaa.yml [hunk 10]
```diff
@@ -508,6 +508,8 @@ uaa:
     url: http://localhost:8080/uaa/approvals
   login:
     url: http://localhost:8080/uaa/authenticate
+  dashboard:
+    uri: http://localhost:8080/dashboard
```
[the remark this answers is not in the document]
Comment (Yogeshkumarb57) [reply uaa/src/main/resources/uaa.yml:512]: Dashboard uri "http://localhost:8080/dashboard", is it hosted by different uaa dashboard app?

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java (hunk 6 above)
Comment (johnson-jeb) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java:72]: I think Mockmvc we are checking the status while fetch the request itself (andExpect(status().isNotFound())), but here we are not checking. So I use this condition to check. Probably I will add the fail assertion in else.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java (hunk 7 above)
Comment (johnson-jeb) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java:93]: I think Mockmvc we are checking the status while fetch the request itself (andExpect(status().isNotFound())), but here we are not checking. So I use this condition to check.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java (hunk 8 above)
Comment (johnson-jeb) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java:105]: I think Mockmvc we are checking the status while fetch the request itself (andExpect(status().isNotFound())), but here we are not checking. So I use this condition to check.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java (hunk 9 above)
Comment (johnson-jeb) [reply uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/ZoneControllerIntegrationTests.java:119]: I think Mockmvc we are checking the status while fetch the request itself (andExpect(status().isNotFound())), but here we are not checking. So I use this condition to check.

File: uaa/src/main/resources/uaa.yml (hunk 10 above)
[the remark this answers is not in the document]
Comment (Yogeshkumarb57) [reply uaa/src/main/resources/uaa.yml:512]: I think "/dashboard" is provided by different app and it should be configurable.

[the remark this answers is not in the document]
Comment (johnson-jeb) [reply uaa/src/main/resources/uaa.yml:512]: Dashboard is a different app will be deployed in the environment, so we cannot form it from current url. So this needs to be configurable.

File: zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneController.java (hunk 5 above)
Comment (johnson-jeb) [reply zone-service/src/main/java/org/cloudfoundry/identity/uaa/zone/ZoneController.java]: Updated

**Review by Yogeshkumarb57 — APPROVED** [review]

**Pull request merged**